pymc_marketing/mmm/budget_optimizer.py

```diff
@@ -98,32 +112,61 @@ def objective(self, budgets: list[float]) -> float:
         )
-        for idx, (_channel, params) in enumerate(self.parameters.items()):
-            budget = budgets[idx] / self.scales[idx]
-            first_params = (
-                params["adstock_params"]
-                if self.adstock_first
-                else params["saturation_params"]
-            )
-            second_params = (
-                params["saturation_params"]
-                if self.adstock_first
-                else params["adstock_params"]
-            )
-            spend = np.full(self.num_periods, budget)
-            spend_extended = np.concatenate([spend, np.zeros(self.adstock.l_max)])
-            transformed_spend = second_transform.function(
-                x=first_transform.function(x=spend_extended, **first_params),
-                **second_params,
-            ).eval()
-            total_response += np.sum(transformed_spend)
-        return -total_response
+
+        budget = budgets / self.scales  # dim: (channels)
+
+        first_params = (
+            self.parameters["adstock_params"]
+            if self.adstock_first
+            else self.parameters["saturation_params"]
+        )
+        second_params = (
+            self.parameters["saturation_params"]
+            if self.adstock_first
+            else self.parameters["adstock_params"]
+        )
+
+        spend = np.tile(budget, (self.num_periods, 1))  # dim: (periods, channels)
+        spend_extended = np.vstack(
+            [spend, np.zeros((self.adstock.l_max, spend.shape[1]))]
+        )  # dim: (periods + l_max, channels)
+
+        _response = first_transform.function(x=spend_extended, **first_params)
+
+        # Dynamically reshape parameters in `second_params` to match `_response`
+        for param_name, param_value in second_params.items():
+            if param_value.ndim == 3:  # Check if it lacks the `time` dimension
+                # Reshape by adding a singleton dimension for broadcasting over time
+                second_params[param_name] = param_value.reshape(
+                    param_value.shape[0], param_value.shape[1], 1, param_value.shape[2]
+                )
+
+        # Apply the second transformation with the reshaped `second_params`
+        response = second_transform.function(x=_response, **second_params).eval()
+        return response.sum(axis=(2, 3)).flatten()  # dim: (chain * draw)
+
+    def objective(self, budgets):
+        """Objective function for the budget optimization."""
+        response_distribution = self._estimate_response(budgets=budgets)
+
+        # Inspect the signature of the objective function
+        sig = inspect.signature(self.objective_function)
+        params = sig.parameters
+
+        # Prepare arguments
+        function_kwargs = self.objective_function_kwargs.copy()
+        if "assets" in params:
+            function_kwargs["assets"] = budgets
+
+        if "capital" in params:
+            function_kwargs["capital"] = np.sum(budgets)
```

Comment:
This isn't scalable. What happened with the Callable[[np.ndarray], float]? Maybe we have `Callable[[np.ndarray, np.ndarray], float]` where the second parameter are the budgets